src/framework-ondata.js

```diff
@@ -1,18 +1,129 @@
 /**
  * File: framework-ondata.js 
  */
+var _ = require('underscore');
+var util = require('util');
+
+var stateFactory = require('./create-state');
+var vniManager = require('./vni-manager');
 
 /** 
- * This is a stub file to allow for a smaller PR and merge.  It will be replaced
- * with the real version of the code once the pipeline-component-factory is finalized.
- *
+ * The standard RDF pipeline framework ondata method. This is called whenever  
+ * there is new data on any port.  This method will update the input state,
+ * and determine if the component updater should be executed now or not. 
+ * If the updater should be executed, the fRunUpdater wrapper API will be
+ * called to to handle that.  The updater output (or error) will be stored 
+ * in output or error state, and those results will be sent on to any 
+ * down stream components for their processing.
+ * 
  * @this port context
- *
- * @param payload data payload on the input port
- * @param socketIndex (optional) this will be set only for ports configured for
- *        multiple inputs (multi/noflo addressable).
+ * @param payload data coming into the port
+ * @param socketIndex socket index if the port is a multiple input (addressable) 
  */
 module.exports = function( payload, socketIndex ) {
 
+     var portName = this.name;
+
+     var vnid = payload.vnid || '';
+     var vni = this.nodeInstance.vni(vnid);
+
+     // Set the input states
+     vni.inputStates( portName, socketIndex, stateFactory( vnid, payload ) );
```

Comment:
These lines should be checking to see if the payload is *already* a state object (as evidenced by the presence of the payload.vnid property, even if its value is the empty string).  A new state object should only be created if the payload is not already a state object.  The reason for this is so that in the normal case, with RDF pipeline nodes connected to RDF pipeline nodes, downstream nodes directly reference the states of upstream nodes, without needing to recreate states.  So lines 27-31 should be:
<pre>
if (!('vnid' in payload)) {
  // Payload is from IIP or a vanilla NoFlo component.
  // Turn it into a state object:
  payload = stateFactory('', payload);
}
var vnid = payload.vnid || '';
var vni = this.nodeInstance.vni(vnid);

// Set the input states
vni.inputStates( portName, socketIndex, payload );
</pre>